arctic_training/data/dpo_factory.py

```diff
@@ -0,0 +1,398 @@
+
+IGNORE_INDEX = -100
+
+class DPODataConfig(DataConfig):
+    max_length: int = 8192
+    """ Maximum length of the input sequence. """
+
+    max_prompt_length: int = 4096
+    """ Maximum prompt length of the input sequence. """
+
+    mask_inputs: bool = True
+    """ Whether to mask the input sequence. """
+
+    always_max_length: bool = False
+    """
+    If this is turned on, each batch will be filled up to the max length by
+    appending samples until the total length matches the max length. It might
+    cause the last sample to be truncated.
+    """
+    dpo_prompt_truncation_mode: str = "keep_start"
+
+def _adjust_prompt_length(
+    prompt_token,
+    chosen_token,
+    rejected_token,
+) -> List[int]:
+    c_len = len(chosen_token["prompt_input_ids"])
+    r_len = len(rejected_token["prompt_input_ids"])
+    min_len = min(c_len, r_len)
+
+    for k, v in prompt_token.items():
+        prompt_token[k] = v[:min_len]
+
+    num_diff_tokens = sum(
+        [
+            a != b
+            for a, b in zip(
+                chosen_token["prompt_input_ids"], rejected_token["prompt_input_ids"]
+            )
+        ]
+    )
+    num_diff_len = abs(c_len - r_len)
+    if num_diff_tokens > 1 or num_diff_len > 1:
+        raise ValueError(
+            "Chosen and rejected prompt_input_ids might only differ on the last token due to tokenizer merge ops."
+        )
+
+    return min_len
+
+
+
+
+def add_bos_token_if_needed(
+    bos_token_id,
+    prompt_len_input_ids,
+    prompt_tokens,
+    chosen_prompt_len_input_ids,
+    chosen_tokens,
+    rejected_prompt_len_input_ids,
+    rejected_tokens,
+) -> Tuple:
+    if bos_token_id is not None:
+        if (
+            prompt_len_input_ids == 0
+            or bos_token_id != prompt_tokens["prompt_input_ids"][0]
+        ):
+            prompt_tokens["prompt_input_ids"] = [bos_token_id] + prompt_tokens[
+                "prompt_input_ids"
+            ]
+            prompt_tokens["prompt_attention_mask"] = [1] + prompt_tokens[
+                "prompt_attention_mask"
+            ]
+        if (
+            chosen_prompt_len_input_ids == 0
+            or bos_token_id != chosen_tokens["prompt_input_ids"][0]
+        ):
+            chosen_tokens["prompt_input_ids"] = [bos_token_id] + chosen_tokens[
+                "prompt_input_ids"
+            ]
+            chosen_tokens["prompt_attention_mask"] = [1] + chosen_tokens[
+                "prompt_attention_mask"
+            ]
+        if (
+            rejected_prompt_len_input_ids == 0
+            or bos_token_id != rejected_tokens["prompt_input_ids"][0]
+        ):
+            rejected_tokens["prompt_input_ids"] = [bos_token_id] + rejected_tokens[
+                "prompt_input_ids"
+            ]
+            rejected_tokens["prompt_attention_mask"] = [1] + rejected_tokens[
+                "prompt_attention_mask"
+            ]
+    return prompt_tokens, chosen_tokens, rejected_tokens
```

Comment:
```suggestion
def add_bos_token_if_needed(
    bos_token_id: Union[None, int],
    tokens: Dict[str, List[int]],
) -> Dict[str, List[int]]:
    len_input_ids = len(tokens["prompt_input_ids"])
    if bos_token_id is not None and (len_input_ids == 0 or bos_token_id != tokens["prompt_token_ids"][0]):
                tokens["prompt_input_ids"].insert(0, bos_token_id)
                tokens["prompt_attention_mask].insert(0, 1)
    return tokens
```

This could be more concise. And then to call the function:
```python
prompt_tokens = add_bos_token_if_needed(tokenizer.bos_token_id, prompt_tokens)
chosen_tokens = add_bos_token_if_needed(tokenizer.bos_token_id, chosen_tokens)
rejected_tokens = add_bos_token_if_needed(tokenizer.bos_token_id, rejected_tokens)
```